src/blob.jl

```diff
@@ -119,23 +119,46 @@ end
 
 # Recursion scales better than splatting for large numbers of fields.
 Base.@assume_effects :foldable @inline function blob_offset(::Type{T}, i::Int) where {T}
-    _recursive_sum_field_sizes(T, Val(i - 1))
+    blob_offsets(T)[i]
 end
 
+Base.@assume_effects :foldable @inline function blob_offsets(::Type{T}) where {T}
+    _recursive_field_offsets(T)
+end
+Base.@assume_effects :foldable _recursive_field_offsets(::Type{T}) where {T} =
+    _recursive_field_offsets(T, Val(fieldcount(T)))
+Base.@assume_effects :foldable _recursive_field_offsets(::Type, ::Val{0}) = ()
+Base.@assume_effects :foldable _recursive_field_offsets(::Type, ::Val{1}) = (0,)
+Base.@assume_effects :foldable function _recursive_field_offsets(::Type{T}, ::Val{i}) where {T,i}
+    tup = _recursive_field_offsets(T, Val(i-1))
+    return (tup..., tup[end] + self_size(fieldtype(T, i-1)))
+end
```

Comment:
For my curiosity/understanding, what is your thinking around :foldable versus force-inlining of these methods? A recomputation consideration? 🤔 (To be clear, all good by me, looking to learn :).)

Reply:
I think in practice there's not much difference, but yeah i was hoping that this might possibly be better for recomputation?

I think I like that this is more explicit about the intent: I don't actually care if the compiler inlines these calls when it's doing its compilation or if it evaluates and caches them. What i do care about is that all of this constant-folds down to a single value, and that value will be inlined at the end. So this feels like a more specific description of my intent? But i'm not quite sure and i don't _think_ there should be any difference.